### ISSUE TITLE
[API Proposal]: Convert DateTime instance to DateOnly or TimeOnly

### ISSUE THREAD

**Issue by mu88** [fluentassertions/fluentassertions#2234]
### Background and motivation

The Fluent Assertion syntax for `DateTime` objects (e. g. `12.April(1953).At(19, 53)`) is phantastic. But IMHO it could be even better with two new extension methods `.ToDateOnly()` and `.ToTimeOnly()`. This would allow keeping a fluent syntax (instead of `new DateOnly(1953, 4, 12)`) and shorter (instead of `DateOnly.FromDateTime(12.April(1953))`) .

### API Proposal

```C#
public static class FluentDateTimeExtensions
{
    public static DateOnly ToDateOnly(this DateTime date);

    public static TimeOnly ToTimeOnly(this DateTime date);
}
```

### API Usage

```C#
var now = 12.April(1953);

var date = now.ToDateOnly();
var time = now.ToTimeOnly();
```

### Alternative Designs

/

### Risks

Since it's only a new extension method to be added, a regression or breaking change should not happen.

**Comment by dennisdoomen** [comment]
I like that. If @jnyrup agrees, would you be willing to provide a PR for this?

**Comment by mu88** [comment]
Yeah sure, feel free to assign the issue to me if both of you consider it worth

**Comment by jnyrup** [comment]
Preamble: I'll poke and challenge the API, it's not to shoot the messenger, but how I get a better grasp and feeling of an API.

I wasn't around when the first date/time extensions were introduced in c143b577974d0edb79fc2a584559dc25146c2c55, but I assume one goal was writing time instants in a more natural language compared to using one of `DateTime`'s many constructors.
```cs
19.April(2020).At(12, 34)
// vs
new DateTime(2020, 4, 19, 12, 34, 00)
```

C# now has target typed new expressions, so one can write a `TimeOnly` quite concise
```cs
timeOfDay.Should().Be(new(13, 37))
```

If one currently wants to use chaining calls to achieve the same result I guess it would be something like this.

```cs
13.Hours(37.Minutes()).ToTimeOnly()
```

Note that the example above assumes we also add this extension method in addition to the proposed ones.
```cs
public static TimeOnly ToTimeOnly(this TimeSpan time);
```

I haven't used `DateOnly` and `TimeOnly` besides the existing tests in FA, so I would appreciate some more complete examples to design an API that fits the usage.
That is, I'd like to avoid creating some API, just "because we can".

**Comment by mu88** [comment]
For a time I could agree with your arguments: `new(13, 37)` is okayish to read, concise, and probably unambiguous.

But for a date writing `new(1953, 4, 12)` doesn't feel good to me, as it could be both April 12 or December 4 if you're not familiar with the API.

**Comment by dennisdoomen** [comment]
> ```cs
> timeOfDay.Should().Be(new(13, 37))
> ```

It might be me, but I would wonder whether this was supposed to mean 13h37 or 13m37s. 

Another example of using `ToTimeOnly` could be:

`19.April(2020).At(12, 34).ToTimeOnly()`

Although I think I would prefer `AsTimeOnly` and `AsDateOnly`

**Comment by jnyrup** [comment]
I'd like to see `19.April(2020).At(12, 34).ToTimeOnly()` used in a more complete example.
I find it a bit odd why one would create an instant in time and then immediately truncate it to the time of day.

**Comment by dennisdoomen** [comment]
> I find it a bit odd why one would create an instant in time and then immediately truncate it to the time of day.

You're totally right. That would be stupid. But `9.Hours().And(30.Minutes()).ToTimeOnly()` does make sense. Same for `9.Hours().And(30.Minutes()).And(45.Seconds()).ToTimeOnly()`

I was initially thinking you may want to truncate some `DateTime` value you are getting from some kind of invocation so that the time part is ignored, but there are already properties on `DateTime` for that. But having `19.April(2020).ToDateOnly()` can be useful on its own _and_ could be used for the truncation part. Same for code that interacts with a database and returns a `DateTime` when you only care about the time part.

**Comment by jnyrup** [comment]
There are some discussions in https://github.com/dotnet/runtime/issues/58734 to add similar APIs directly to BCL.

**Comment by mu88** [comment]
And to add another nuance: my initial motivation was just `ToDateOnly()` and I just added `ToTimeOnly()` for completeness.

**Comment by dennisdoomen** [comment]
> There are some discussions in [dotnet/runtime#58734](https://github.com/dotnet/runtime/issues/58734) to add similar APIs directly to BCL.

That would be the best solution, but right not it's an API suggestion.

**Comment by dennisdoomen** [comment]
I still vote for a yes.

**Comment by jnyrup** [comment]
> Although I think I would prefer `AsTimeOnly` and `AsDateOnly`

As we're converting another type and truncating, I vote for using the `To` prefix.

https://stackoverflow.com/questions/13867829/to-vs-as-vs-get-method-prefixes

**Comment by dennisdoomen** [comment]
So I guess that's a _yes_.

**Comment by mu88** [comment]
Cool, I will take care about this after my vacation

**Comment by Corniel** [comment]
What about this?

Instead of creating a `DateTime` with `11.June(2017)`, create a custom `DateBuilder` that can both implictly cast to `DateOnly` as to ``DateTime`?

The contract of such a builder could like like this:

``` C#
namespace FluentAssertions.Extensions;

public readonly struct DateBuilder : IEquatable<DateBuilder>
{
    public DateBuilder(int year, int month, int day);

    public DateTime At(TimeSpan time);

    public DateTime At(int hours, int minutes, int seconds = 0, int milliseconds = 0, int microseconds = 0, int nanoseconds = 0);
    
	public DateTimeOffset WithOffset(TimeSpan offset);
    
	public DateTimeOffset WithOffset(int hours, int minutes = 0, int seconds = 0);

    public static DateTime operator -(DateBuilder builder, TimeSpan time) => builder.At(-time);

    public static DateTime operator +(DateBuilder builder, TimeSpan time) => builder.At(+time);

    public static implicit operator DateTime(DateBuilder builder) => builder.date;

#if NET6_0_OR_GREATER
    public static implicit operator DateOnly(DateBuilder builder) => DateOnly.FromDateTime(builder.date);
#endif
}
```

**Comment by dennisdoomen** [comment]
> The contract of such a builder could like like this:

Can you add some usage examples?

**Comment by Corniel** [comment]
Sure

**Comment by dennisdoomen** [comment]
I actually like that what you did @Corniel. If @jnyrup agrees, I suggest we take that path.

**Comment by IT-VBFK** [comment]
Hmm.. IMHO
`DateTimeOffset dateTime = 10.December(2011).Offset(0).At(09, 30, 45, 123, 456, 700);` doesn't read that fluent anymore.

**Comment by Corniel** [comment]
@IT-VBFK Renaming to `.WithOffset()` will do the trick?

**Comment by IT-VBFK** [comment]
Have edited my comment :)

**Comment by Corniel** [comment]
@IT-VBFK The problem with returning `DateOnly` is that is does not implicitly casts to `DateTime`, so `DateTime dt = 11.June(2017);` will not compile anymore.

Furthermore, `DateOnly` is not available in older versions of .NET, so that we would have to add a package for older versions.

For me, the most important is the implicit cast, the rest is bonus to me.

**Comment by IT-VBFK** [comment]
@jnyrup Can we polyfill this?

**Comment by IT-VBFK** [comment]
> The problem with returning DateOnly is that is does not implicitly casts to DateTime, so DateTime dt = 11.June(2017); will not compile anymore.

Ok, what about adding the `WithOffset(x)` to the end (after `At()`)? And if it is possible assuming offset = 0 when omitted?

**Comment by dennisdoomen** [comment]
`WithOffset` could actually change the type of the builder to something like `DateTimeOffsetBuilder`.

**Comment by jnyrup** [comment]
> > Furthermore, DateOnly is not available in older versions of .NET, so that we would have to add a package for older versions.
> 
> @jnyrup Can we polyfill this?

Polyfill `DateOnly`?
I can't imagine that. For anyone to use it, they would have to depend on FA, and FA is not designed for that.

**Comment by Corniel** [comment]
If `DateOnly` does not exist, there is no use for it, a test input, is it? Furthermore, It would require extra chaining for the `DateTime` scenario, I think that is undesirable.

@dennisdoomen We can choose to have a builder from the start 'till the end, in that case the builder needs an extra (private) `TimeSpan` offset, and een implicit cast to `DateTimeOffset`. I personally think, that only having this builder for the first step (`11.June(2017)`) should be sufficient. But I can live with with both impementations..

**Comment by dennisdoomen** [comment]
You still want to be able to chain `At(12,0)` to a call to `10.March(2012).WithOffset(1.Hours())`.

**Comment by Corniel** [comment]
@dennisdoomen That can be achieved both by an extension on `DateTimeOffset`, or by having a builder. The downside of keeping the builder is - now I think of it again you can have this:

``` C#
DateTime dt = 11.Juni(2017).At(06, 15).WithOffset(1.Hours());
```
And this case, should the offfset than be applied?

**Comment by dennisdoomen** [comment]
> And this case, should the offfset than be applied?

I guess so. It shouldn't matter where you first call `WithOffset` and then `At` or not

**Comment by Corniel** [comment]
It should both be possible, and it should be both possible, also when the builder is only for the first step.

**Comment by Corniel** [comment]
I've updated my implementation. For Offsets if now have both `.AsOffset()`, and `.WithOffset(timespan)`. But in this potentially contradicts the  `As` v.s. `To` debate.

``` C#
var dateTime = 10.March(2012).AsOffset();
var otherDateTime = 10.Juni(2016).WithOffset(1.Hours()).At(06, 15);
```

**Comment by Corniel** [comment]
I've made some changes so that the `DateBuilder` itself can also be used directly (in anonymous object etc.) To do assertions.

**Comment by dennisdoomen** [comment]
> I've made some changes so that the `DateBuilder` itself can also be used directly (in anonymous object etc.) To do assertions.

How did you solve the usage of the fluent API when constructing an anonymous objects?

**Comment by Corniel** [comment]
> How did you solve the usage of the fluent API when constructing an anonymous objects?

By implementing equality on both `DateTime` and `DateOnly`. Unfortunately, I also had to extend the `ObjectExtensions.GetComparer()` because the default Equality comparer checks for a.Equals(b), but not b.Equals(a), but I think that is fair price to pay, as other scenario's also potentially benefit from this.

See: https://github.com/fluentassertions/fluentassertions/pull/2279/files

**Comment by dennisdoomen** [comment]
You mean you changed the implementation of 'BeEquivalentTo' so it understands how to compare an instance of 'DateBuilder' to a 'DateOnly'/'TimeOnly'? If so, that's quite neat. 

However, I'm getting the feeling this solution is becoming a bit much for what we want to accomplish. What do you think @jnyrup ?

**Comment by jnyrup** [comment]
In any case `GetComparer` should not be touched here.

**Comment by Corniel** [comment]
> In any case `GetComparer` should not be touched here.

Not at all, or not in this PR? In general, I think its preferable that it checked if A.Equals(B) or B.Equals(A).

**Comment by jnyrup** [comment]
I'm not opposing to ever changing it, but we'll need to have a better understanding of the consequences of relaxing this equality check to compensate for non-symmetric `Equals` methods.
I'd like to have that analysis kept separate from this PR, it simply helps me tracking the work better (right now I have ~25 FA issues/PRs/discussions to dig trough).

**Comment by mu88** [comment]
Am I correct that the originally intended approach of `ToTimeOnly()` and `ToDateOnly()` is no longer in favor?

**Comment by dennisdoomen** [comment]
> Am I correct that the originally intended approach of `ToTimeOnly()` and `ToDateOnly()` is no longer in favor?

It is still my preferred approach, but what @Corniel is suggesting sounds appealing. But we haven't settled on the implementation details.  

> Not at all, or not in this PR? In general, I think its preferable that it checked if A.Equals(B) or B.Equals(A).

@Corniel what if we also introduce dedicated implementations of `IEquivalencyStep` that know how to compare a `DateBuilder` on the expectation to an instance of `DateTIme`,`DateOnly` and `TimeOnly` on the subject?

**Comment by mu88** [comment]
Thx @dennisdoomen 👍🏻 so I'll just wait a little bit and continue following the conversation

**Comment by Corniel** [comment]
> @Corniel what if we also introduce dedicated implementations of `IEquivalencyStep` that know how to compare a `DateBuilder` on the expectation to an instance of `DateTIme`,`DateOnly` and `TimeOnly` on the subject?

I see no objects to that. I can give that a try.
